pyth/client/src/client_cache.rs

```diff
@@ -0,0 +1,189 @@
+use {
+    crate::{client::PythClientTrait, error, PythClient},
+    async_stream::stream,
+    async_trait::async_trait,
+    grug::{Binary, Inner, Lengthy, NonEmpty},
+    indexer_disk_saver::{error::Error, persistence::DiskPersistence},
+    reqwest::{IntoUrl, Url},
+    std::{
+        collections::HashMap,
+        env,
+        path::{Path, PathBuf},
+        sync::{
+            atomic::{AtomicU64, Ordering},
+            Arc,
+        },
+        thread::sleep,
+        time::Duration,
+    },
+    tokio::runtime::Runtime,
+    tokio_stream::StreamExt,
+};
+
+#[derive(Debug, Clone)]
+pub struct PythClientCache {
+    base_url: Url,
+    // Used to return newer vaas at each call.
+    vaas_index: Arc<AtomicU64>,
+}
+
+impl PythClientCache {
+    pub fn new<U: IntoUrl>(base_url: U) -> Result<Self, error::Error> {
+        Ok(Self {
+            base_url: base_url.into_url()?,
+            vaas_index: Arc::new(AtomicU64::new(0)),
+        })
+    }
+
+    /// Load data from cache or retrieve it from the source.
+    pub fn load_or_retrieve_data<I>(
+        base_url: Url,
+        ids: NonEmpty<I>,
+    ) -> HashMap<PathBuf, Vec<Vec<Binary>>>
+    where
+        I: IntoIterator + Lengthy + Clone,
+        I::Item: ToString,
+    {
+        let mut stored_vaas = HashMap::new();
+
+        // Load data for each id.
+        for id in ids.into_inner() {
+            let element = id.to_string();
+            let filename = Self::cache_filename(&element);
+
+            #[allow(clippy::map_entry)]
+            // If the file is not in memory, try to read from disk.
+            if !stored_vaas.contains_key(&filename) {
+                let cache_file = DiskPersistence::new(filename.clone(), true);
+
+                // If the file does not exists, retrieve the data from the source.
+                if !cache_file.exists() {
+                    let rt = Runtime::new().unwrap();
+                    let values = rt.block_on(async {
+                        let client = PythClient::new(base_url.clone()).unwrap();
+
+                        let mut stream = client
+                            .stream(NonEmpty::new(vec![id.to_string()]).unwrap())
+                            .await
+                            .unwrap();
+
+                        let mut values = vec![];
+                        while values.len() < 15 {
```

Comment:
Maybe we should put that `15` in a static at the top of the file to easily change that if needed.